clang/tools/clang-linker-wrapper/SYCLOffloadWrapper.cpp

```diff
@@ -0,0 +1,748 @@
+    SmallVector<Constant *> PropSetsInits;
+    for (const auto &PropSet : *PropRegistry) {
+      // create content in the rightmost column and get begin/end pointers
+      std::pair<Constant *, Constant *> Props =
+          addPropertySetToModule(PropSet.second);
+      // get the next the middle column element
+      auto *Category = addStringToModule(PropSet.first, "SYCL_PropSetName");
+      PropSetsInits.push_back(ConstantStruct::get(SyclPropSetTy, Category,
+                                                  Props.first, Props.second));
+    }
+    // now get content for the leftmost column - create the top-level
+    // PropertySetsBegin/PropertySetsBegin entries and return pointers to them
+    return addStructArrayToModule(PropSetsInits, SyclPropSetTy);
+  }
+
+  // Emits an object that holds <address, size> pair for the device image
+  // and puts it into a .tgtimg section. This section can be used for
+  // finding and extracting all device images from the fat binary after
+  // linking.
+  void emitRegistrationFunctions(Constant *Address, size_t Size, size_t ImageID,
+                                 StringRef OffloadKindTag) {
+    Type *IntPtrTy = M.getDataLayout().getIntPtrType(C);
+    auto *ImgInfoArr =
+        ConstantArray::get(ArrayType::get(IntPtrTy, 2),
+                           {ConstantExpr::getPointerCast(Address, IntPtrTy),
+                            ConstantInt::get(IntPtrTy, Size)});
+    auto *ImgInfoVar = new GlobalVariable(
+        M, ImgInfoArr->getType(), true, GlobalVariable::InternalLinkage,
+        ImgInfoArr, Twine(OffloadKindTag) + Twine(ImageID) + Twine(".info"));
+    ImgInfoVar->setAlignment(
+        MaybeAlign(M.getDataLayout().getTypeStoreSize(IntPtrTy) * 2u));
+    ImgInfoVar->setUnnamedAddr(GlobalValue::UnnamedAddr::Local);
+    ImgInfoVar->setSection(".tgtimg");
+
+    // Add image info to the used list to force it to be emitted to the
+    // object.
+    appendToUsed(M, ImgInfoVar);
+  }
+
+  Constant *wrapImage(const SYCLImage &Image, size_t ImageID,
+                      StringRef OffloadKindTag) {
+    auto *NullPtr = Constant::getNullValue(PointerType::getUnqual(C));
+    // DeviceImageStructVersion change log:
+    // -- version 2: updated to PI 1.2 binary image format
+    constexpr uint16_t DeviceImageStructVersion = 2;
+    constexpr uint8_t SYCLOffloadKind = 4;
+    auto *Version =
+        ConstantInt::get(Type::getInt16Ty(C), DeviceImageStructVersion);
+    auto *Kind = ConstantInt::get(Type::getInt8Ty(C), SYCLOffloadKind);
+    auto *Format = ConstantInt::get(
+        Type::getInt8Ty(C),
+        static_cast<int8_t>(binaryImageFormatToInt8(Image.Format)));
+    auto *Target =
+        addStringToModule(Image.Target, Twine(OffloadKindTag) +
+                                            Twine("target.") + Twine(ImageID));
+    auto *CompileOptions = addStringToModule(
+        Image.CompileOptions,
+        Twine(OffloadKindTag) + Twine("opts.compile.") + Twine(ImageID));
```

Comment:
I think that you should be able to omit second and third explicit constructions of `Twine`, because it should have `operator+` overloads for `const char *` and `StringRef`